### PR TITLE
fix: Fix merge into bug

### LINKED ISSUE #13732: bug: Merge Into Panic
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

DatabendQuery v1.2.207-nightly-410f3dae9f(rust-1.75.0-nightly-2023-11-12T22:08:26.586848221Z)

### What's Wrong?

Merge Into Panic because of type mismatch

### How to Reproduce?

```sql
-- GPT4 generated SQL
create database mergeinto;
use mergeinto;




CREATE TABLE orders CLUSTER BY (to_yyyymmddhh(created_at), user_id) AS SELECT
    number % 500000 AS order_id,
    number % 100000 AS user_id,
    CASE WHEN (rand() * 10)::int % 2 = 0 THEN 'buy'
         ELSE 'sell'
    END AS order_type,
    CASE WHEN (rand() * 10)::int % 3 = 0 THEN 'BTC'
         WHEN (rand() * 10)::int % 3 = 1 THEN 'ETH'
         ELSE 'XRP'
    END AS asset_type,
    (rand() * 100)::decimal(18, 8) AS quantity,
    (rand() * 1000)::decimal(18, 8) AS price,
    CASE WHEN (rand() * 10)::int % 3 = 0 THEN 'completed'
         WHEN (rand() * 10)::int % 3 = 1 THEN 'pending'
         ELSE 'cancelled'
    END AS status,
    date_add('day', floor(rand() * 10 % 365)::int, '2021-01-01') AS created_at,
    date_add('day', floor(rand() * 10 % 365)::int, '2021-01-01') AS updated_at
FROM numbers(500000);



set global enable_experimental_merge_into = 1;
-- merge into
MERGE INTO orders USING
(
    SELECT
        CASE 
            WHEN number % 2 = 0 THEN (number / 2) % 250000 
            ELSE number % 500000 + 500000 
        END AS order_id,
        number % 100000 AS user_id,
        CASE WHEN (rand() * 10)::int % 2 = 0 THEN 'buy'
             ELSE 'sell'
        END AS order_type,
        CASE WHEN (rand() * 10)::int % 3 = 0 THEN 'BTC'
             WHEN (rand() * 10)::int % 3 = 1 THEN 'ETH'
             ELSE 'XRP'
        END AS asset_type,
        (rand() * 100)::decimal(18, 8) AS quantity,
        (rand() * 1000)::decimal(18, 8) AS price,
        CASE WHEN (rand() * 10)::int % 3 = 0 THEN 'completed'
             WHEN (rand() * 10)::int % 3 = 1 THEN 'pending'
             ELSE 'cancelled'
        END AS status,
        date_add('day', floor(rand() * 10 % 365)::int, '2021-01-01') AS created_at,
        date_add('day', floor(rand() * 10 % 365)::int, '2021-01-01') AS updated_at
    FROM numbers(500000)
) AS source
ON orders.order_id = source.order_id
WHEN MATCHED THEN 
    UPDATE SET
        orders.user_id = source.user_id,
        orders.order_type = source.order_type,
        orders.asset_type = source.asset_type,
        orders.quantity = source.quantity,
        orders.price = source.price,
        orders.status = source.status,
        orders.created_at = source.created_at,
        orders.updated_at = source.updated_at
WHEN NOT MATCHED THEN 
    INSERT (order_id, user_id, order_type, asset_type, quantity, price, status, created_at, updated_at)
    VALUES (source.order_id, source.user_id, source.order_type, source.asset_type, source.quantity, source.price, source.status, source.created_at, source.updated_at);
```
```text
ERROR 1105 (HY000): PanicError. Code: 1104, Text = assertion `left == right` failed: column data type mismatch at index: 0, expr: order_id
  left: Nullable(Number(UInt32))
 right: Number(UInt32).
 ```
### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by JackTan25**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR
fix #13732 #13733
- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13744)

### CODE REVIEW

File: src/query/sql/src/planner/optimizer/optimizer.rs [hunk 1]
```diff
@@ -127,14 +127,28 @@ pub fn optimize(
...
+                plan.meta_data.clone(),
+                plan.input.child(1)?.clone(),
+            )?;
+            // replace right source
+            let mut join_sexpr = plan.input.clone();
+            join_sexpr = Box::new(join_sexpr.replace_children(vec![
+                Arc::new(join_sexpr.child(0)?.clone()),
+                Arc::new(right_source),
+            ]));
+            // Box::new(optimize_query(ctx, opt_ctx, metadata.clone(), *s_expr)?),
```
Comment (BohuTANG) [inline src/query/sql/src/planner/optimizer/optimizer.rs:143]:
```suggestion
          
```

File: src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs [hunk 2]
```diff
@@ -318,8 +319,39 @@ impl Processor for MatchedSplitProcessor {
...
                 metrics_inc_merge_into_append_blocks_counter(1);
                 metrics_inc_merge_into_append_blocks_rows_counter(current_block.num_rows() as u32);
+
+                // cornor case: for merge into update, if the target table's column is not null,
+                // for example, target table has three columns like (a,b,c), and we use update set target_table.a = xxx,
+                // it's fine because we have cast the xxx'data_type into a's data_type in `generate_update_list()`,
+                // but for b,c, the hash table will transform the origin data_type (b_type,c_type) into
+                // (nullable(b_type),nullable(c_type)), so we will get datatype not match error, let's transform
+                // them back here.
+                let res_entries = current_block.columns();
```
Comment (BohuTANG) [inline src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs:329]: @b41sh need your help to review this cast codes.

File: src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs [hunk 3]
```diff
@@ -318,8 +319,39 @@ impl Processor for MatchedSplitProcessor {
...
+                            data_type: res_entries[idx].data_type.clone(),
+                            display_name: "".to_string(),
+                        }),
+                        dest_type: field.data_type().clone(),
+                    })
+                }
+                let cast_operator = BlockOperator::Map {
+                    exprs: cast_exprs,
+                    projections: Some((res_entries.len()..res_entries.len() * 2).collect()),
+                };
```
Comment (BohuTANG) [inline src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs:348]:
```suggestion
                let current_columns = current_block.columns();
                assert_eq!(
                    self.target_table_schema.fields.len(),
                    current_columns.len(),
                    "target table columns and current columns length mismatch"
                );

                let cast_exprs = current_columns
                    .iter()
                    .enumerate()
                    .map(|(idx, col)| Expr::Cast {
                        span: None,
                        is_try: false,
                        expr: Box::new(Expr::ColumnRef {
                            span: None,
                            id: idx,
                            data_type: col.data_type.clone(),
                            display_name: "".to_string(),
                        }),
                        dest_type: self.target_table_schema.fields[idx].data_type().clone(),
                    })
                    .collect::<Vec<_>>();

                let cast_operator = BlockOperator::Map {
                    exprs: cast_exprs,
                    projections: Some((current_columns.len()..current_columns.len() * 2).collect()),
                };
```

File: src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs [hunk 4]
```diff
@@ -318,8 +319,44 @@ impl Processor for MatchedSplitProcessor {
                 current_block = op.execute(&self.ctx.get_function_context()?, current_block)?;
                 metrics_inc_merge_into_append_blocks_counter(1);
                 metrics_inc_merge_into_append_blocks_rows_counter(current_block.num_rows() as u32);
+
+                // cornor case: for merge into update, if the target table's column is not null,
```
Comment (BohuTANG) [inline src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs:323]: Prefer spill these codes to a new function, like:
```
normalize_column_types_for_merge(ctx, ...) -> Result<Datablock>
```
To make the process logic more clear and more modular.

Comment (JackTan25) [reply src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs:323]: ok

File: src/query/sql/src/planner/optimizer/optimizer.rs [hunk 5]
```diff
@@ -127,14 +127,28 @@ pub fn optimize(
...
         }
         Plan::MergeInto(plan) => {
+            // optimize source :fix issue #13733
+            let right_source = optimize_query(
+                ctx.clone(),
+                opt_ctx.clone(),
+                plan.meta_data.clone(),
+                plan.input.child(1)?.clone(),
+            )?;
+            // replace right source
```
Comment (b41sh) [inline src/query/sql/src/planner/optimizer/optimizer.rs:137]: Why are we replacing the right source? better to add some comments to explain the reason.

Comment (JackTan25) [reply src/query/sql/src/planner/optimizer/optimizer.rs:137]: for example the sql in https://github.com/datafuselabs/databend/issues/13733, there is subquery expr like `CASE 
            WHEN number % 2 = 0 THEN (number / 2) % 250000
            ELSE (SELECT MAX(order_id) FROM orders) + number + 1
        END AS order_id,`, so we will get invalid index in https://github.com/datafuselabs/databend/blob/main/src/query/sql/src/planner/semantic/lowering.rs#L234 ,the same with https://github.com/datafuselabs/databend/blob/main/src/query/sql/src/planner/semantic/lowering.rs#L205,https://github.com/datafuselabs/databend/blob/main/src/query/sql/src/planner/semantic/lowering.rs#L180,etc. So we will use optimizer to solve the index

File: src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs (hunk 2 above)
Comment (b41sh) [reply src/query/storages/fuse/src/operations/merge_into/processors/processor_merge_into_matched_and_split.rs]: This cast seems a bit redundant, but there's no better way to do it at the moment, so we can implement it this way first and see if there's a better way to modify it later.

**Review by b41sh — APPROVED** [review]

**Pull request merged**